### PR TITLE
[Elastic Agent] Fix issues with dynamic inputs and conditions

### LINKED ISSUE #23685: [Agent] Dynamic inputs fail to get validated
It was found during https://github.com/elastic/beats/pull/23679. Dynamic inputs setup to automatically discover `scheduler` and `controllermanager` Pods does not work right now and we get the following error:
```
2021-01-25T15:36:29.224Z	DEBUG	application/periodic.go:40	Failed to read configuration, error: could not emit configuration: could not create the AST from the configuration: missing field accessing 'inputs' (source:'/etc/agent.yml')
```

Agent version:
```
Daemon: 7.11.0-SNAPSHOT (build: 5691b7c654663ee51f1d347ce02bda332fb300e7 at 2021-01-25 09:48:16 +0000 UTC)
Binary: 7.11.0-SNAPSHOT (build: 5691b7c654663ee51f1d347ce02bda332fb300e7 at 2021-01-25 09:48:16 +0000 UTC)
```

`agent.yml`:
```yml
    id: ef9cc740-5bf0-11eb-8b51-39775155c3f5
    revision: 2
    outputs:
      default:
        type: elasticsearch
        hosts:
          - >-
            ${ES_HOST}
        username: ${ES_USERNAME}
        password: ${ES_PASSWORD}
    agent:
      monitoring:
        enabled: true
        use_output: default
        logs: true
        metrics: true
    providers.kubernetes:
      node: ${NODE_NAME}
      scope: node
    inputs:
      - id: 934ef8aa-ed19-405b-8160-ebf62e3d32f8
        name: kubernetes-node-metrics
        revision: 1
        type: kubernetes/metrics
        use_output: default
        meta:
          package:
            name: kubernetes
            version: 0.2.8
        data_stream:
          namespace: default
        streams:
          - id: >-
              kubernetes/metrics-kubernetes.controllermanager-3d50c483-2327-40e7-b3e5-d877d4763fe1
            data_stream:
              dataset: kubernetes.controllermanager
              type: metrics
            metricsets:
              - controllermanager
            hosts:
              - '${kubernetes.pod.ip}:10252'
            period: 10s
            #condition: startsWith(${kubernetes.pod.name}, "kube-controller-manager")
            condition: ${kubernetes.pod.labels.component} == 'kube-controller-manager'
          - id: >-
              kubernetes/metrics-kubernetes.scheduler-3d50c483-2327-40e7-b3e5-d877d4763fe1
            data_stream:
              dataset: kubernetes.scheduler
              type: metrics
            metricsets:
              - scheduler
            hosts:
              - '${kubernetes.pod.ip}:10251'
            period: 10s
            #condition: startsWith(${kubernetes.pod.name}, "kube-scheduler")
            condition: ${kubernetes.pod.labels.component} == 'kube-scheduler'
          - id: >-
              kubernetes/metrics-kubernetes.proxy-3d50c483-2327-40e7-b3e5-d877d4763fe1
            data_stream:
              dataset: kubernetes.proxy
              type: metrics
            metricsets:
              - proxy
            hosts:
              - 'localhost:10249'
            period: 10s
          - id: >-
              kubernetes/metrics-kubernetes.container-934ef8aa-ed19-405b-8160-ebf62e3d32f8
            data_stream:
              dataset: kubernetes.container
              type: metrics
            metricsets:
              - container
            add_metadata: true
            bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
            hosts:
              - 'https://${env.NODE_NAME}:10250'
            period: 10s
            ssl.verification_mode: none
```

Converting `${NODE_NAME}` placeholders to `${env.NODE_NAME}` does not fix the problem and even if we remove all other datastream configs and leave only the dynamic one it still gives the error.

In addition, if we remove the dynamic inputs part and have `${env.NODE_NAME}` we still get the same error.

Also a condition like `condition: startsWith(${kubernetes.pod.name}, "kube-controller-manager")` fails and the `inspect` command gives the following:

```
condition "startsWith(${kubernetes.pod.name}, "kube-controller-manager")" evaluation failed: startsWith: accepts exactly 2 string arguments; recieved *eql.null and string
```

Related to https://github.com/elastic/beats/issues/23613

### LINKED ISSUE #23829: [Ingest Manager]: Data Stream not generated for standalone Agent on kibana.
**Kibana version:**
7.11 BC5 Kibana Cloud environment 

**Host OS and Browser version:**
Windows 10, All

**Preconditions**
1.  7.11 BC5 Kibana Cloud environment should be available.
```
BUILD 37827
COMMIT d801a7bb08e17368584d00fbd97a5d0006285b51
Artifact link: https://staging.elastic.co/7.11.0-903dc0b6/downloads/beats/elastic-agent/elastic-agent-7.11.0-windows-x86_64.zip
```
2. Default policy having system integration should exist.[No change in namespace].

**Steps to reproduce:**
1. Login to Kibana cloud environment.
2. Go to Add Agents flyout. Go To 'Run Standalone' tab.
3. Download elastic-agent.yml And update username/password with kibana cloud credentials.
4. RDP to endpoint. Replace elastic-agent.yml at extracted elastic-agent folder.
5. Open Powershell <run ./elastic-agent.exe install command.
6. Say no to enroll to fleet.

**Expected Result:**
Data Stream should be available for standalone Agent on kibana.

**Logs:**
[elastic-agent-json.log](https://github.com/elastic/beats/files/5917050/elastic-agent-json.log)
[elastic-agent-watcher-json.log](https://github.com/elastic/beats/files/5917055/elastic-agent-watcher-json.log)

**Screenshots:** 

image

**However on running './elastic-agent version' command, Below error is generated:**

image

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Fixes 2 issues that where affecting the usage of dynamic inputs and conditions.

* Loading of the configuration was not excluding the `inputs` from the go-ucfg variable expansion. Inside of `inputs` on variable expansion is done by the dynamic inputs code.
* EQL string methods where too strict on type checks, this lowers that to try to convert any type to a string before doing the comparison. This is important in cases where a variable results to an un-expected type, like `*eql.null`.

## Why is it important?

Usage of dynamic inputs work as expected.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #23829
- Closes #23685

### CODE REVIEW

File: x-pack/elastic-agent/pkg/config/config.go [hunk 1]
```diff
@@ -47,26 +37,42 @@ func NewConfigFrom(from interface{}, opts ...ucfg.Option) (*Config, error) {
...
 		}
-		c, err := yaml.NewConfig(content, opts...)
+	} else {
+		c, err := ucfg.NewFrom(from, opts...)
 		return newConfigFrom(c), err
 	}
 
-	c, err := ucfg.NewFrom(from, opts...)
-	return newConfigFrom(c), err
+	var c map[string]interface{}
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/config/config.go:59]: can we generalize this a bit? e.g have a list of keywords where we want to skip ucfg resolution. 
i think extraction to func and var list of keys would be enough

**Review by michalpristas — APPROVED** [review]
tested ok with standalone config, code lgtm

**Comment by blakerouse** [comment]
@michalpristas I cleaned up the config loading more. See what you think.

File: x-pack/elastic-agent/pkg/config/config.go [hunk 2]
```diff
@@ -32,10 +50,26 @@ func New() *Config {
...
+			ucfgOpts = append(ucfgOpts, ot)
+		case Option:
+			localOpts = append(localOpts, ot)
+		default:
+			return nil, fmt.Errorf("unknown option type %T", o)
+		}
+	}
+	local := &options{}
+	for _, o := range localOpts {
+		o(local)
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/config/config.go:71]: with multiple VarSkipKeys last one wins right? should we allow multiple entries by combining them?

**Review by michalpristas — APPROVED** [review]
I think it looks ok. i like having load defines inside config package

File: x-pack/elastic-agent/pkg/config/config.go (hunk 2 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/config/config.go:71]: Yes last wins. If we combine them, then no way to overwrite.... idk its a choice

Comment (michalpristas) [reply x-pack/elastic-agent/pkg/config/config.go:71]: i'm bringing this up just to verify it's a decision not a bug

**Comment by ChrsMark** [comment]
Thank you for fixing this one @blakerouse . Could you provide an example of valid configuration based on your fixes (dynamic input + env vars), or point me to a docs page which includes such examples? It's still not clear to me what is the proper way to combine dynamic inputs and env vars, do you still need `env.` prefix when refer to env variables in configuration like at 
```
data_stream:
  dataset: kubernetes.container
  type: metrics
  metricsets:
  - container
  add_metadata: true
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  hosts:
  - 'https://${env.NODE_NAME}:10250'
```
?

**Comment by blakerouse** [comment]
@ChrsMark Yes you still need to use `${env.NODE_NODE}`. Inside of inputs is only resolved by the providers so it must start with the name of the provider first. So `env.*` in the case you want to get it from the environment variable.

**Comment by ChrsMark** [comment]
Hey, is there anything missing with this so as to have it in and unblock https://github.com/elastic/beats/pull/23679?

**Comment by blakerouse** [comment]
Tested locally with same built package and both Fleet mode and standalone work. So E2E test failed for an unknown reason.

**Pull request merged**